base/src/main/java/org/mozilla/jss/pkcs7/EncryptedContentInfo.java

```diff
@@ -236,28 +238,59 @@ public static EncryptedContentInfo createCRSCompatibleEncryptedContentInfo(OBJEC
         }
 
         // get the key gen parameters
-        AlgorithmIdentifier algid = contentEncryptionAlgorithm;
-        KeyGenAlgorithm kgAlg = KeyGenAlgorithm.fromOID( algid.getOID() );
+        KeyGenAlgorithm kgAlg = KeyGenAlgorithm.fromOID( contentEncryptionAlgorithm.getOID() );
         if( !(kgAlg instanceof PBEAlgorithm) ) {
             throw new NoSuchAlgorithmException("KeyGenAlgorithm is not a"+
                 " PBE algorithm");
         }
-        ASN1Value params = algid.getParameters();
+        ASN1Value params = contentEncryptionAlgorithm.getParameters();
         if( params == null ) {
             throw new InvalidAlgorithmParameterException(
                 "PBE algorithms require parameters");
         }
-        PBEParameter pbeParams;
-        if( params instanceof PBEParameter) {
-            pbeParams = (PBEParameter) params;
-        } else {
+        byte[] salt = null;
+        int iterations = 0;
+        EncryptionAlgorithm encAlg = null;
+        AlgorithmParameterSpec algParams = null;
+        HMACAlgorithm hashAlg = null;
+        if(!kgAlg.toOID().equals(PBEAlgorithm.PBE_PKCS5_PBES2.toOID())) {
+            PBEParameter pbeParams;
+            if( params instanceof PBEParameter) {
+                pbeParams = (PBEParameter) params;
+            } else {
+                byte[] encodedParams = ASN1Util.encode(params);
+                pbeParams = (PBEParameter)
+                    ASN1Util.decode( PBEParameter.getTemplate(), encodedParams );
+            }
+            salt = pbeParams.getSalt();
+            iterations = pbeParams.getIterations();
+            encAlg = ((PBEAlgorithm)kgAlg).getEncryptionAlg();
+        }
+        else {
             byte[] encodedParams = ASN1Util.encode(params);
-            pbeParams = (PBEParameter)
-                ASN1Util.decode( PBEParameter.getTemplate(), encodedParams );
+            PBES2Params pbe2Params = (PBES2Params)
```

Comment:
According to the specification `params` should be present and if not we cannot go ahead. If they are not present the pk12 is not valid. We could have a custom exception for p12 not valid in this case but I am not sure if it is something to consider at this level.

Reply:
Testing with OpenSSL p12 files there could be problem. I am investigating them.

Reply:
The problem is only with `KeyLength` which is optional for the specification and `OpenSSL` was not adding. The overall `params` cannot be null because the `encode` or `decode` method would throw an exception.

Reply:
OK.